src/common/KokkosKernels_Sorting.hpp

```diff
@@ -0,0 +1,631 @@
+    Ordinal jump = boxSize / 2;
+    Comparator comp;
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(t, work),
+      [=](const Ordinal i)
+      {
+        Ordinal elem1 = workStart + i;
+        Ordinal elem2 = workStart + jump + i;
+        if(elem2 < Ordinal(v.extent(0)))
+        {
+          if(comp(v(elem2), v(elem1)))
+          {
+            Value temp = v(elem1);
+            v(elem1) = v(elem2);
+            v(elem2) = temp;
+          }
+        }
+      });
+    if(teamsPerBox == 1)
+    {
+      //This team can finish phase 2 for all the smaller red boxes that follow,
+      //since there are no longer cross-team data dependencies
+      for(Ordinal subLevel = 1; subLevel < logBoxSize; subLevel++)
+      {
+        t.team_barrier();
+        Ordinal logSubBoxSize = logBoxSize - subLevel;
+        Ordinal subBoxSize = Ordinal(1) << logSubBoxSize;
+        Kokkos::parallel_for(Kokkos::TeamThreadRange(t, work),
+          [=](const Ordinal i)
+          {
+            Ordinal globalThread = i + t.league_rank() * work;
+            Ordinal subBox = globalThread >> (logSubBoxSize - 1);
+            Ordinal subBoxStart = subBox << logSubBoxSize;
+            Ordinal subBoxOffset = globalThread & ((Ordinal(1) << (logSubBoxSize - 1)) - 1); //i % (subBoxSize / 2)
+            Ordinal elem1 = subBoxStart + subBoxOffset;
+            //later phases (pink box): within a block, compare with fixed distance (boxSize / 2) apart
+            Ordinal elem2 = elem1 + subBoxSize / 2;
+            if(elem2 < Ordinal(v.extent(0)))
+            {
+              if(comp(v(elem2), v(elem1)))
+              {
+                Value temp = v(elem1);
+                v(elem1) = v(elem2);
+                v(elem2) = temp;
+              }
+            }
+          });
+      }
+    }
+  };
+  View v;
+  Ordinal boxSize;
+  Ordinal teamsPerBox;
+};
+
+//Version to be called from host on a single array
+//Generally ~2x slower than Kokkos::sort() for large arrays (> 50 M elements),
+//but faster for smaller arrays.
+//
+//This is more general than BinSort: bitonic supports any trivially copyable type
+//and an arbitrary device-compatible comparison operator (provided through operator() of Comparator)
+//If comparator is void, use operator< (which should only be used for primitives)
+template<typename View, typename ExecSpace, typename Ordinal, typename Comparator = DefaultComparator<typename View::value_type>>
+void bitonicSort(View v)
+{
+  typedef Kokkos::TeamPolicy<ExecSpace> team_policy;
+  typedef typename team_policy::member_type team_member;
+  Ordinal n = v.extent(0);
+  //If n is small, just sort on a single team
+  if(n <= Ordinal(1) << 16)
+  {
+    Kokkos::parallel_for(team_policy(1, Kokkos::AUTO()),
+        BitonicSingleTeamFunctor<View, Ordinal, team_member, Comparator>(v));
+  }
+  else
+  {
+    //Partition the data equally among fixed number of teams
+    const Ordinal numTeams = 256;
```

Comment:
Should this be some functon of n and the hardware ?

Reply:
I just tried to pick a power-of-two value that is greater than the number of SMs on any existing GPU, so that this function will get full occupancy. Does Kokkos have a way to get the number of SMs? With that I could compute this at runtime. Although, I don't think it will have a big effect on performance since CUDA will schedule multiple blocks onto one SM if there is register space (I think).

Reply:
@srajama1 I just checked and Kokkos::Cuda::concurrency() returns the maximum concurrent number of threads over all SMs. Kokkos::Cuda does provide the SM count through ``cuda_internal_multiprocessor_count()``. I'll look at how other KK functions choose team size and then do something based on that.